File: Chapter.01/QuickSort.FSharp/Benchmark.fs
namespace QuickSort.FSharp

module Benchmark =

    open System
    open BenchmarkDotNet.Attributes
 
    [<MemoryDiagnoser>]
    [<RPlotExporter; RankColumn>]
    type BenchmarkQuickSort() =

        [<Params(1000, 10000)>]
        [<DefaultValue>]
        val mutable N : int
        
        [<DefaultValue>]
        val mutable iterations : int[]
        
        [<GlobalSetup>]
        member this.Setup() =
            let rand = new Random((int) DateTime.Now.Ticks)
            this.iterations <- Array.init this.N (fun _ -> rand.Next())
            
        [<Benchmark>]    
        member this.Sequential() = Functions.quicksortSequential (this.iterations |> Array.toList)    

        [<Benchmark>]    
        member this.Parallel() = Functions.quicksortParallel (this.iterations |> Array.toList)    
        
        [<Benchmark>]    
        member this.ParallelDepth() = Functions.quicksortParallelWithDepth Functions.ParallelismHelpers.MaxDepth (this.iterations |> Array.toList)

File: Chapter.01/QuickSort.CSharp/Benchmark.cs
using System;
using System.Linq;
using BenchmarkDotNet.Attributes;

namespace QuickSort.CSharp
{
    [MemoryDiagnoser]
    [RPlotExporter]
    [RankColumn]
    public class BenchmarkQuickSort
    {
        public int[] iterations;
        [Params(1000, 10000)] public int N;

        [GlobalSetup]
        public void Setup()
        {
            var rand = new Random((int) DateTime.Now.Ticks);
            iterations = Enumerable.Range(0, N).Select(_ => rand.Next()).ToArray();
        }

        [Benchmark]
        public void Sequential()
        {
            Functions.QuickSort_Sequential(iterations);
        }

        [Benchmark]
        public void Parallel()
        {
            Functions.QuickSort_Parallel(iterations);
        }

        [Benchmark]
        public void ParallelDepth()
        {
            Functions.QuickSort_Parallel_Threshold(iterations);
        }
    }
}